pandas/core/groupby.py

```diff
@@ -1329,6 +1329,27 @@ def cumsum(self, axis=0):
 
     @Substitution(name='groupby')
     @Appender(_doc_template)
+    def rank(self, axis=0, method='average', numeric_only=True,
+             na_option='keep', ascending=True, pct=False):
+        """Compute numerical data ranks (1 through n) along axis.
+        """
+
+        if numeric_only:
+            data = self._obj_with_exclusions._get_numeric_data()
+            if data.size == 0:
+                raise DataError('No numeric types to aggregate')
+            data = data.groupby(self.grouper)
+        else:
```

Comment:
I'm not sure I understand what you are saying. If I remove this section, so that the method becomes
```
    def rank(self, axis=0, method='average', numeric_only=True,
             na_option='keep', ascending=True, pct=False):

        data = self

        def wrapper(values):
            return values.rank(axis=axis, method=method, na_option=na_option,
                               ascending=ascending, pct=pct)

        return data.transform(wrapper)
```
then I get:
```
ERROR: test_rank (pandas.tests.test_groupby.TestGroupBy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/nicolas/Projects/pandas/pandas/tests/test_groupby.py", line 3603, in test_rank
    self.assertRaises(DataError, dg.rank, method='first')
  File "/home/nicolas/Projects/pandas/pandas/util/testing.py", line 2300, in assertRaises
    _callable(*args, **kwargs)
  File "/home/nicolas/Projects/pandas/pandas/core/groupby.py", line 1372, in rank
    return data.transform(wrapper)
  File "/home/nicolas/Projects/pandas/pandas/core/groupby.py", line 3496, in transform
    return self._transform_general(func, *args, **kwargs)
  File "/home/nicolas/Projects/pandas/pandas/core/groupby.py", line 3444, in _transform_general
    raise ValueError(msg)
ValueError: transform must return a scalar value for each group
```

If instead I define the method as:
```
    def rank(self, axis=0, method='average', numeric_only=True,
             na_option='keep', ascending=True, pct=False):

        if numeric_only:
            data = self._obj_with_exclusions._get_numeric_data()
            if data.size == 0:
                raise DataError('No numeric types to aggregate')

        def wrapper(values):
            return values.rank(axis=axis, method=method, na_option=na_option,
                               ascending=ascending, pct=pct)

        return self.transform(wrapper)
```
then I get
```
  File "/home/nicolas/Projects/pandas/pandas/tests/test_groupby.py", line 3612, in test_rank
    result = df.groupby('c').rank(method='first')
  File "/home/nicolas/Projects/pandas/pandas/core/groupby.py", line 1369, in rank
    return self.transform(wrapper)
  File "/home/nicolas/Projects/pandas/pandas/core/groupby.py", line 3493, in transform
    return self._transform_general(func, *args, **kwargs)
  File "/home/nicolas/Projects/pandas/pandas/core/groupby.py", line 3441, in _transform_general
    raise ValueError(msg)
ValueError: transform must return a scalar value for each group
```

What do you mean when you say the flag is handled at a higher level?